### ISSUE TITLE
Disable accessibility alerts during the ResetAllButton listener

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/scenery-phet#291]
When the user presses the ResetAllButton, all alerts should be disabled to prevent a stream of accessibility alerts that could be produced as the sim is reset.  Instead, the user should just hear a single summary of the alert action.

**Comment by jessegreenberg** [comment]
The above commits handle this by wrapping the ResetAllButton listener in a function that does the following in order:
- disables all alerts
- calls the ResetAllButtonListener
- enables all alerts
- Announces that the whole screen was reset

@jbphet would you please review these changes? Since AriaHerald is outfitted for PhET-iO you can see the behavior in the "console" wrapper if you wish.

**Comment by jbphet** [comment]
I think there may be a problem with this.  The way the listener is attached is by eventually calling the inherited function `RoundPushButton.addListener`.  So, while I don't think anyone is doing this, it would be possible to create a ResetAllButton and then call `addListener` later, and it would bypass what you're trying to accomplish here.  Consider instead overriding `addListener` instead of wrapping the listener in the constructor.

If for some reason this approach is infeasible, here is a comment to consider if the wrapping remains in the constructor: I found it a little odd that the listener is wrapped regardless of whether it is defined.  It works with no defined listener (I tried it by commenting out the listener set in the scenery-phet demo), so it's not a functional problem, but I think it would be easier to understand the code if it doesn't wrap the listener if it isn't defined.

I have another observation that isn't directly related to the changes made above, but looking at the code brought this up.  AriaHerald has a `setDisabled` function.  In general, I think we try to set an enabled value rather than a disabled value because it avoids the whole 'double negative' thing.  I would highly recommend changing this to be a `setEnabled` (and perhaps add an ES5 getter setter).

And finally, this may seem picky, but the text of the reset message is  'Sim screen restarted, everything reset'.  The word 'restarted' kind of sounds like the sim was reloaded or something, and the listener would have to interpret both the words 'restarted' and 'reset'.  If it were me hearing this, I would wonder what the distinction was between the two.  Can this be simplified to something like 'simulation screen has been reset' or 'sim screen reset?

**Comment by jessegreenberg** [comment]
@jbphet thanks for the review!

> I think there may be a problem with this...it would be possible to create a ResetAllButton and then call addListener later, and it would bypass what you're trying to accomplish here.

That is a good point, this was not handled previously.  The commit above should fix this.

> The way the listener is attached is by eventually calling the inherited function RoundPushButton.addListener.

That is true, *except* for the listener added through options, which is added directly to ```PushButtonModel``` without going through ```addListener```. So in https://github.com/phetsims/scenery-phet/commit/0bb107fe09d0883155304511acd283e5fc0401b0 I am removing the listener from ResetAllButton options, and adding it in the constructor with an ```addListener``` function that overrides ```RoundPushButton```.

Fixes for the other three comments have also been done in https://github.com/phetsims/scenery-phet/commit/0bb107fe09d0883155304511acd283e5fc0401b0. @jbphet could you please review again?

**Comment by jbphet** [comment]
The changes for making AriaHerald have a methods for setting `enabled` rather than `disabled` look good, as does the update message.

The fix for wrapping the listener seems a bit awkward to me, but it's a result of the way the button hierarchy works.  I'm going to get some input from the other developers to decide whether we should support making it possible to override `addListener` and have it apply to the listener supplied in the options.

**Comment by terracoda** [comment]
@jessegreenberg and @jbphet, @emily-phet and I discussed the change to the _Reset All_ alert. I have started a new issue https://github.com/phetsims/a11y-research/issues/21 to discuss this change to determine the best wording.

**Comment by jessegreenberg** [comment]
In https://github.com/phetsims/a11y-research/issues/21, we decided to change the alert back.

**Comment by jbphet** [comment]
While working on https://github.com/phetsims/phet-io-wrapper-sonification/issues/4, I noticed that there is a problem with the solution that has been committed for this issue.  The overridden `addListener` function that is in `ResetAllButton` as of this writing looks like this:

```js
addListener: function( listener ) {
      var self = this;

      // wrap the listener in a function that disables alerts until the listener
      // returns - then announce that the Screen has been reset 
      var accessibleListener = function() {
        self.startedCallbacksForResetEmitter.emit();

        AriaHerald.callWithDisabledAlerts( listener );
        AriaHerald.announcePolite( resetAllAlertString );

        self.endedCallbacksForResetEmitter.emit();
      };

      // @private - add the accessibility listener so button fires on 'enter' or 'spacebar'
      this.clickListener = this.addAccessibleInputListener( { click: accessibleListener } );

      ResetButton.prototype.addListener.call( this, accessibleListener );
    }
```

This is going to make the Aria Herald announcement and emit the started and ended callbacks for every listener that is added to the ResetAllButton.  In most cases, there will only be once listener, but we can by no means count on this.  A better solution would be to override the "fire" function, which invokes all listeners.

**Comment by jbphet** [comment]
Assigning to @jessegreenberg and @zepumph to discuss and work on a solution.

**Comment by zepumph** [comment]
Here is PushButtonModel.fire():
```js
    /**
     * Fires all listeners.
     * @public (phet-io, a11y)
     */
    fire: function() {
      this.startedCallbacksForFiredEmitter.emit();
      var copy = this.listeners.slice( 0 );
      copy.forEach( function( listener ) {
        listener();
      } );
      this.endedCallbacksForFiredEmitter.emit();
    }
```

It would be nice to set up the override in such a way, that we make sure that these emitters and listeners are still called.

**Comment by jessegreenberg** [comment]
@zepumph and I worked on this and changed the strategy.  Rather than wrapping each listener with special functionality (which suffers from https://github.com/phetsims/scenery-phet/issues/291#issuecomment-300618644), we are using ```startedCallbacksForFiredEmitter``` and ```endedCallbacksForFiredEmitter``` to disable alerts when callbacks start, and enable alerts when callbacks end.  I think this is the same approach sonification is using to prevent sounds from coming through while properties are reset.

@jbphet could you please review https://github.com/phetsims/scenery-phet/commit/3f38bba9e53e193de9580d52b3a2ebfb55bb64e3?

**Comment by zepumph** [comment]
@samreid I would also love to hear from you on 3f38bba. Are we ok with using this technically 'phet-io specific' emitter for this accessibility solution. It's possible that we shouldn't be tapping into them like this, meaning (1) from a child (2) non phet-io code.

**Comment by samreid** [comment]
I think it's generally OK to use the `endedCallbacksForFiredEmitter` and similar callbacks--I suspected they may have utility outside of PhET-iO.

**Comment by jbphet** [comment]
The changes look reasonable to me, and as long as @samreid approves of the usage of these emitters that were originally added in support of phet-io, I'm okay with it.  Closing.
